zcash_client_sqlite/src/wallet/init/migrations/full_account_ids.rs

```diff
@@ -0,0 +1,381 @@
+
+use super::{add_account_birthdays, receiving_key_scopes, v_transactions_note_uniqueness};
+
+/// The migration that switched from presumed seed-derived account IDs to supporting
+/// HD accounts and all sorts of imported keys.
+pub(super) const MIGRATION_ID: Uuid = Uuid::from_u128(0x1b104345_f27e_42da_a9e3_1de22694da43);
+
+pub(crate) struct Migration {
+    pub(super) seed: Rc<Option<SecretVec<u8>>>,
+}
+
+impl schemer::Migration for Migration {
+    fn id(&self) -> Uuid {
+        MIGRATION_ID
+    }
+
+    fn dependencies(&self) -> HashSet<Uuid> {
+        [
+            receiving_key_scopes::MIGRATION_ID,
+            add_account_birthdays::MIGRATION_ID,
+            v_transactions_note_uniqueness::MIGRATION_ID,
+        ]
+        .into_iter()
+        .collect()
+    }
+
+    fn description(&self) -> &'static str {
+        "Replaces the `account` column in the `accounts` table with columns to support all kinds of account and key types."
+    }
+}
+
+impl RusqliteMigration for Migration {
+    type Error = WalletMigrationError;
+
+    fn up(&self, transaction: &Transaction) -> Result<(), WalletMigrationError> {
+        transaction.execute_batch(
+            r#"
+            PRAGMA foreign_keys = OFF;
+            PRAGMA legacy_alter_table = ON;
+
+            CREATE TABLE accounts_new (
+                id INTEGER NOT NULL PRIMARY KEY AUTOINCREMENT,
+                account_type INTEGER NOT NULL DEFAULT 0,
+                hd_seed_fingerprint BLOB,
+                hd_account_index INTEGER,
+                uvk TEXT NOT NULL,
+                birthday_height INTEGER NOT NULL,
+                recover_until_height INTEGER,
+                CHECK (
+                    (account_type = 0 AND hd_seed_fingerprint IS NOT NULL AND hd_account_index IS NOT NULL)
+                    OR
+                    (account_type != 0 AND hd_seed_fingerprint IS NULL AND hd_account_index IS NULL)
+                )
+            );
+            CREATE UNIQUE INDEX accounts_uvk ON accounts_new ("uvk");
+            "#,
+        )?;
+
+        // We require the seed *if* there are existing accounts in the table.
+        if transaction.query_row("SELECT COUNT(*) FROM accounts", [], |row| {
+            Ok(row.get::<_, u32>(0)? > 0)
+        })? {
+            if let Some(seed) = &self.seed.as_ref() {
+                let seed_id = HdSeedFingerprint::from_seed(seed);
+                // Although 'id' is an AUTOINCREMENT column, we'll set it explicitly to match the old account value
+                // strictly as a matter of convenience to make this migration script easier,
+                // specifically around updating tables with foreign keys to this one.
+                transaction.execute(r#"
+                INSERT INTO accounts_new (id, account_type, hd_seed_fingerprint, hd_account_index, uvk, birthday_height, recover_until_height)
+                SELECT account, 0, :seed_id, account, ufvk, birthday_height, recover_until_height
```

Comment:
Change the 0 to not be a magic number.

Reply:
This came up in another comment as well. Magic numbers are our _friends_ when it comes to migrations. Because if the constant in code were ever to change, we'd need another db migration, and _this_ migration would need to continue setting to 0 rather than the new number. We can't have migrations silently changing behavior over time.